test/e2e/common.go

```diff
@@ -359,3 +363,106 @@ func IsBootedFromDisk(client *ssh.Client) (bool, error) {
 
 	return isDisk, nil
 }
+
+func DeploymentRolledOut(ctx context.Context, clusterProxy framework.ClusterProxy, name string, namespace string, desiredGeneration int64) bool {
+	clientSet := clusterProxy.GetClientSet()
+	deploy, err := clientSet.AppsV1().Deployments(namespace).Get(ctx, name, metav1.GetOptions{})
+	Expect(err).To(BeNil())
+	if deploy != nil {
+		// When the number of replicas is equal to the number of available and updated
+		// replicas, we know that only "new" pods are running. When we also
+		// have the desired number of replicas and a new enough generation, we
+		// know that the rollout is complete.
+		return (deploy.Status.UpdatedReplicas == *deploy.Spec.Replicas) &&
+			(deploy.Status.AvailableReplicas == *deploy.Spec.Replicas) &&
+			(deploy.Status.Replicas == *deploy.Spec.Replicas) &&
+			(deploy.Status.ObservedGeneration >= desiredGeneration)
+	}
+	return false
+}
+
+func CopyFile(src, dst string) error {
+	out, err := os.Create(filepath.Clean(dst))
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = out.Close()
+	}()
+
+	in, err := os.Open(filepath.Clean(src))
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = in.Close()
+	}()
+
+	_, err = io.Copy(out, in)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type InstallIronicBMOInput struct {
+	Kustomization       string
+	IronicEnvFile       string
+	ClusterProxy        framework.ClusterProxy
+	WaitForDeployment   bool
+	WatchDeploymentLogs bool
+	DeploymentName      string
+	DeployNS            string
+	LogPath             string
+	WaitIntervals       []interface{}
+}
+
+func InstallIronicBMO(ctx context.Context, input *InstallIronicBMOInput) error {
```

Comment:
I don't like this function. It does strange things without explanation (copying files). It is very specific (ironic/bmo) instead of generic (e.g. apply and wait for deployment). There is no doc string.
Don't get me wrong, I can see the usefulness of a function like this, but it should then be generic and well documented.

Reply:
My bad. The design and naming of this function is affected by the one in CAPM3 e2e, so I had an impression that it's a common design choice and needn't to be explained further. I'll add some docs.

Other than that reason, I thought of making more generic function, but as we've already a function to install cert manager, I think having one for BMO and Ironic is not so bad a design? This allows us to add some specific logic only needed for our usecase, and after all, this is meant to serve our e2e tests, so it doesn't need to be so generic it can be reused elsewhere.

Reply:
Alright, I get it. Unfortunately CAPM3 is not a good "role model", so please don't take too much inspiration from it :sweat_smile: Especially everything related to how we deploy Ironic/BMO in there is just bad because it is heavily integrated and influenced by metal3-dev-env and the `deploy.sh` script. We should not reimplement that but rather try to learn from those mistakes and make something better here.

With that out of way, I think the only issue here is really the handling of that env file. It is "premature optimization" and not needed. If you just get rid of that, this function can be completely generic, useful and nice! :slightly_smiling_face: